gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,342 @@
+    private static final int NUM_TRIES_FETCH_TOPIC = 3;
+    private static final int NUM_TRIES_FETCH_OFFSET = 3;
+
+    private final Map<String, SimpleConsumer> activeConsumers = Maps.newHashMap();
+
+    @Override
+    public List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist) {
+      List<TopicMetadata> topicMetadataList = getFilteredMetadataList(blacklist, whitelist);
+
+      List<KafkaTopic> filteredTopics = Lists.newArrayList();
+      for (TopicMetadata topicMetadata : topicMetadataList) {
+        List<KafkaPartition> partitions = getPartitionsForTopic(topicMetadata);
+        filteredTopics.add(new KafkaTopic(topicMetadata.topic(), partitions));
+      }
+      return filteredTopics;
+    }
+
+    private List<KafkaPartition> getPartitionsForTopic(TopicMetadata topicMetadata) {
+      List<KafkaPartition> partitions = Lists.newArrayList();
+
+      for (PartitionMetadata partitionMetadata : topicMetadata.partitionsMetadata()) {
+        partitions.add(new KafkaPartition.Builder().withId(partitionMetadata.partitionId())
+            .withTopicName(topicMetadata.topic()).withLeaderId(partitionMetadata.leader().id())
+            .withLeaderHost(partitionMetadata.leader().host()).withLeaderPort(partitionMetadata.leader().port())
+            .build());
+      }
+      return partitions;
+    }
+
+    private List<TopicMetadata> getFilteredMetadataList(Set<String> blacklist, Set<String> whitelist) {
+      List<TopicMetadata> filteredTopicMetadataList = Lists.newArrayList();
+
+      //Try all brokers one by one, until successfully retrieved topic metadata (topicMetadataList is non-null)
+      for (String broker : KafkaWrapper.this.brokers) {
+        filteredTopicMetadataList = fetchTopicMetadataFromBroker(broker, blacklist, whitelist);
+        if (filteredTopicMetadataList != null) {
+          break;
+        }
+      }
+
+      if (filteredTopicMetadataList == null) {
+        throw new RuntimeException(
+            "Fetching topic metadata from all brokers failed. See log warning for more information.");
+      }
+      return filteredTopicMetadataList;
+    }
+
+    private List<TopicMetadata> fetchTopicMetadataFromBroker(String broker, Set<String> blacklist, Set<String> whitelist) {
+
+      List<TopicMetadata> topicMetadataList = fetchTopicMetadataFromBroker(broker);
+      List<TopicMetadata> filteredTopicMetadataList = Lists.newArrayList();
+      for (TopicMetadata topicMetadata : topicMetadataList) {
+        if (MessageBasedSource.survived(topicMetadata.topic(), blacklist, whitelist)) {
+          filteredTopicMetadataList.add(topicMetadata);
+        }
+      }
+      return filteredTopicMetadataList;
+    }
+
+    private List<TopicMetadata> fetchTopicMetadataFromBroker(String broker, String... selectedTopics) {
+      LOG.info(String.format("Fetching topic metadata from broker %s", broker));
+      SimpleConsumer consumer = null;
```

Comment:
The Kafka Scala API can throw undeclared checked exceptions (which is why Camus needs to use "catch (Exception e)" in many places). Creating consumer is one of those places.